libraries/AP_Scripting/examples/quadruped.lua

```diff
@@ -0,0 +1,308 @@
+local gait_lift_divisor = 0 -- when a leg is lifted and brought back down the action is divided into 2 or multiple steps, so the travel distance also need to be split in between the steps to make the transition natural
+local gait_half_lift_height = 0 -- used to split lift across two steps
+local gait_speed_divisor = 0    -- number of steps in the gait the leg is touching the floor, this is used as a factor to split the travel distance between the steps
+
+local gait_pos_x = {0,0,0,0}    -- X-axis position for each leg (front-right, front-left, back-right, back-left)
+local gait_pos_y = {0,0,0,0}    -- Y-axis position for each leg (front-right, front-left, back-right, back-left)
+local gait_pos_z = {0,0,0,0}    -- Z-axis position for each leg (front-right, front-left, back-right, back-left)
+local gait_rot_z = {0,0,0,0}    -- Z-axis rotation for each leg (front-right, front-left, back-right, back-left)
+local leg_index = 0
+local last_angle = {0,0,0,0,0,0,0,0,0,0,0,0}
+local current = {0,0,0,0,0,0,0,0,0,0,0,0}
+local start_time = 0
+local curr_target = 0
+local pwm = {1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500, 1500}
+
+function Gaitselect()
+    if (gait_type == 0) then
+        -- alternating gait
+        gait_step_total = 6
+        gait_step_leg_start = {1,4,1,4}
+        gait_lifted_steps = 2
+        gait_down_steps = 1
+        gait_lift_divisor = 2
+        gait_half_lift_height = 1
+        gait_speed_divisor = 4
+    elseif (gait_type == 1) then
+        -- wave gait with 28 steps
+        gait_step_total = 28
+        gait_step_leg_start = {8,15,1,22}
+        gait_lifted_steps = 3
+        gait_down_steps = 2
+        gait_lift_divisor = 2
+        gait_half_lift_height = 3
+        gait_speed_divisor = 24
+    end
+end
+
+-- Calculate Gait sequence
+function calc_gait_sequence()
+    local move_requested = (math.abs(x_speed) > speed_dz) or (math.abs(y_speed) > speed_dz) or (math.abs(yaw_speed) > speed_dz)
+
+    if move_requested then
+        for leg_index=1, 4 do
+            update_leg(leg_index)
+        end
+
+        gait_step = gait_step + 1
+        if (gait_step>gait_step_total) then
+            gait_step = 1
+        end
+    else
+        gait_pos_x = {0,0,0,0}
+        gait_pos_y = {0,0,0,0}
+        gait_pos_z = {0,0,0,0}
+        gait_rot_z = {0,0,0,0}
+    end
+end
+
+-- in order for the robot to move forward it needs to move its legs in a
+-- specific order and this is repeated over and over to attain linear motion. when a
+-- specific leg number is passed the update_leg() produces the set of values for the
+-- given leg at that step, for each cycle of the gait each leg will move to a set
+-- distance which is decided by the x_speed, yaw_speed, y_speed
+function update_leg(moving_leg)
+    local leg_step = gait_step - gait_step_leg_start[moving_leg]
+    local move_requested = (math.abs(x_speed) > speed_dz) or (math.abs(y_speed) > speed_dz) or (math.abs(yaw_speed) > speed_dz)
```

Comment:
Isn't this duplicating the `move_requested` calculation already done up in `calc_gait_sequence`, why not pass `move_requested` as an argument to this function and not recalculate it for each leg?

Reply:
yes, good idea.  I've added this to the list of issues to correct with the script. https://github.com/ArduPilot/ardupilot/issues/15112.  So we won't fix this as part of this PR but we have added it to the to-do list.